### ISSUE TITLE
Features

### ISSUE THREAD

**Issue by dannye** [dannye/crystal-tracker#1]
####
- [x] Song Viewer
  - [x] File parsing
    - [x] Support modern pokecrystal macro syntax
    - [x] Bonus: Handle TCG-compatibility macros used by pokered-crysaudio
    - [x] Validate song header
    - [x] Detect invalid macros/arguments
    - [x] Detect nested loops/calls
  - [x] Render notes for all channels
    - [x] Visualize settings changes with colored flags
    - [x] Visualize loops and calls with wrapper boxes
    - [x] Visualize unreferenced labels
    - [x] Visualize tempo changes
  - [x] Customizable timeline
    - [x] Multiple zoom levels
    - [x] Key labels
    - [x] Note labels
    - [x] Configurable Measure ruler
    - [x] Adjustable grid line spacing
    - [x] Theme picker
    - [x] Full screen

####
- [x] Song Player
  - [x] Real-time playback with reasonably accurate tempo, volume, vibrato, etc.
  - [x] Process custom channel 3 wave_samples.asm
  - [x] Process custom channel 4 drumkits.asm
  - [x] Start/resume from any point
  - [x] Loop infinitely
  - [x] Detect when subsequent iterations of the main loop are not exactly identical to the first
  - [x] Mute individual channels during playback
  - [x] Stereo/Mono output
  - [x] Bookmarks
  - [ ] Support remaining miscellaneous commands: pitch_sweep, duty_cycle_pattern, pitch_offset

####
- [x] Song Editor
  - [x] New Song dialog with customizable options
  - [x] Channel selector (only the selected channel may be edited)
  - [x] Multi-note selection
  - [x] Rectangle select
  - [x] Basic note operations (move, resize, put, delete)
  - [x] Edit note properties (speed, volume, vibrato, etc.)
  - [x] Pencil Mode (draw and erase notes)
  - [x] Format Painter (copy note properties from one note to another)
  - [x] Undo/Redo
  - [x] Resize song
  - [x] Make and unmake loops and calls
    - [x] Loops: Reduce, Extend, Unroll, Create
    - [x] Calls: Delete, Unpack, Create, Insert
  - [ ] Add/Remove channels

####
- [x] Extras
  - [x] Channel 3 wave editor
  - [ ] Channel 4 drumkit editor
  - [ ] MIDI device input
  - [ ] MIDI file import

**Comment by nehochupechatat** [comment]
What do you think of implementing a General MIDI export?
With instrumentation similar to Joao Buaes' MIDIs

**Comment by dannye** [comment]
> What do you think of implementing a General MIDI export?

It isn't really useful and doesn't really serve the purpose of this editor. The primary goal of this editor is to directly edit/export *.asm song files for immediate use in a pokecrystal-based project. So that you can press "Save" in Crystal Tracker and immediately run `make` on your project and instantly have a new ROM file with your new song modifications included.

MIDI editors exist for creating MIDI files.

**Comment by nehochupechatat** [comment]
Ah, I understand. What about implementing a "traditional tracker" view? I.e OpenMPT, Furnace, etc
Some musicians prefer to work that way + your tracker is already based off OpenMPT

**Comment by dannye** [comment]
There are several problems with trying to imagine this working with a traditional tracker view. The Pokemon song data format is not well suited for it.

Most fundamentally is that in Pokemon, the song channels are _completely_ independent from each other. One channel can loop a small pattern many times while another channel plays a long sequence of all unique notes. It's also possible for a channel's entire body to be much shorter than another channel's body (ie, the short channel may fully repeat many times in the time that it takes the longer channel to play once. In fact, the channels are so independent that there is no requirement that all the channels loop back to the start at the same point in time -- or that they even loop back to the same point at the beginning.

So the concept of a "row" can't really apply here in the way that it would need to for a traditional tracker view to be effective.

**Comment by duckiedev** [comment]
Would it be possible to add an option under the View menu that can be toggled on and off to display the note + octave as shown in the mock-up below?
![image](https://github.com/user-attachments/assets/3766ad4f-3b0d-4506-9366-7fecff972454)

**Comment by dannye** [comment]
@duckiedev It would be possible but my preference is to keep the list of toggles pretty modest, and I prefer the current behavior of just pitches over pitches+octaves.

**Comment by dotmatrixgirl** [comment]
Trackers such as Klystrack and Goattracker 2 support different pattern length/count per channel, so that block could be overcome.

**Comment by dannye** [comment]
I'm afraid you've missed the point. I'm saying the entire concept of "patterns", the "order", and a "row" are not applicable here.

The channels each have their own independent speed ("ticks per row"), which can change arbitrarily. They each have independent "loops", and "subroutines", and "loops nested inside of subroutines", and "subroutines nested inside of loops", and independent total durations, and independent "starting points" to loop back to at the end.

They have no clear and obvious points of separation between so-called "patterns". Each channel is just a massive sequence of commands+notes.

The channels are so utterly desynced from each other that many, many sacrifices would have to be made to create anything resembling a tracker view; to the point that it would be barely any better than a text editor. The amount of effort to create a GUI for such a view is very obviously not worth it.

You could imagine an *.asm song file that is so well structured that it uses a bunch of subroutines (the "patterns") and uses a long list of calls (the "order") to arrange the subroutines, but this editor seeks to support all valid pokecrystal *.asm song files.

**Comment by PiplupRox-Productions** [comment]
Greetings, I'm a friend of Porygondolier's. Thanks to the recent update I'm finally back to work on using Crystal Tracker to work on music for a project Porygondolier and I have been working on for the past year, and I was wondering: Would it be possible to make it so that, when you add a new note, it plays that note audibly for a moment like when you click it on the piano bar? I have trouble matching the notes to the rows sometimes, since I'm used to programs like Famistudio where if I misplace a note I know immediately because it plays the note I've added. Not a necessary feature by any means, but I would certainly appreciate its inclusion if possible.

**Comment by dannye** [comment]
> Would it be possible to make it so that, when you add a new note, it plays that note audibly for a moment like when you click it on the piano bar?

@PiplupRox-Productions That's a good idea. I was already planning on doing something similar when I add MIDI device input for the next version. (ie, when you press keys on an external MIDI device while in "record" mode, play the note audibly in the editor.)
I'll add an option to do the same thing for "pencil" mode too.

**Comment by PiplupRox-Productions** [comment]
Awesome!
By the way, Porygondolier says he's very glad you're still actively working on this, and I have to say, I am too. It's a great tool, and I'm finally getting comfortable with using it. Keep up the great work :)
There _is_ one other thing. When I use multiple note speeds in a given channel, I often have to place a note a few times with my mouse in slightly different spots X-wise to get the note to align temporally. Is there a way around that or is that just an inherent part of this medium? Porygondolier has suggested fiddling with the ruler settings but I haven't quite figured out how that works yet.

**Comment by dannye** [comment]
> When I use multiple note speeds in a given channel, I often have to place a note a few times with my mouse in slightly different spots X-wise to get the note to align temporally. Is there a way around that or is that just an inherent part of this medium?

@PiplupRox-Productions Yeah, "messy rests" is Crystal Tracker's biggest pain point. The rests are invisible but they each retain their "speed" which can get messier the more the song is edited. If the alignment it off, then it can prevent you from placing a new note in exactly the position that you want.

If you end up with a span of messy rests, the easiest way to fix it is:
Simply use the pencil to drag a super long note across the rests then simply delete the note. **The entire rest now in its place will have the same speed as the note that was just deleted.**

You can change the speed of the note before deleting it, if desired.

The ruler won't help because it's purely visual.

Now, all that being said, the new version of Crystal Tracker that was just released has a change that improves this issue:
The "Speed" input (in the Note Properties panel) is now a dropdown instead of a "stepper" (the number input with the little up and down buttons).
This greatly improves the "messy rest" problem because changing the speed an excessive number of times will tend to create a bunch of tiny rests of odd speeds, and the stepper made it easy for example to go from 12->11->10->9->8.
With the dropdown, you just change the speed directly from 12->8 and then the messy rests are kept at a minimum.

Hope this helps!

**Comment by PiplupRox-Productions** [comment]
Thanks, I'll try that!
